client.d.ts

```diff
@@ -0,0 +1,316 @@
+   * @warning this option is not supported by Firefox.
+   * @default false
+   */
+  serializedOctets: boolean
+
+  /**
+   * Set to `true` to send using buffer and `false` to transmits files a base 64-encoded strings.
+   * @default true
+   */
+  useBuffer: boolean
+
+  /**
+   * The size of the file "chunks" to be loaded at a time.
+   *
+   * Setting this parameter to 0 disables chunking of files.
+   * @default 102400 (100kb)
+   */
+  chunkSize: number
+
+  /**
+   * Specify the topic to listen on.
+   * Need to be the same that the one specified in the client.
+   * @default siofu
+   */
+  topicName: string
+
+/**
+   * WrapData allow you to wrap the Siofu messages into a predefined format.
+   * You can then easily use Siofu packages even in strongly typed topic.
+   * wrapData can be a boolean or an object. It is false by default.
+   * If wrapData is true it will allow you to send all the messages to only one topic by wrapping the siofu actions and messages.
+   *
+   * ```js
+   * {
+   *   action: 'complete',
+   *   message: {
+   *     id: id,
+   *     success: success,
+   *     detail: fileInfo.clientDetail
+   *   }
+   * }
+   * ```
+   *
+   * If wrapData is an object constituted of two mandatory key and one optional:
+   * wrapKey and unwrapKey (mandatory): Corresponding to the key used to wrap the siofu data and message
+   * additionalData (optional): Corresponding to the data to send along with file data
+   *
+   * ```js
+   * if wrapData = {
+   *   wrapKey: {
+   *     action: 'actionType',
+   *     message: 'data'
+   *   },
+   *   unwrapKey: {
+   *     action: 'actionType',
+   *     message: 'message'
+   *   },
+   *   additionalData: {
+   *     acknowledgement: true
+   *   }
+   * }
+   * ```
+   *
+   * When Siofu will send for example a complete message this will send:
+   * ```js
+   * {
+   *   acknowledgement: true,
+   *   actionType: 'complete',
+   *   data: {
+   *     id: id,
+   *     success: success,
+   *     detail: fileInfo.clientDetail
+   *   }
+   * }
+   * ```
+   *
+   * and it's waiting from client data formatted like this:
+   * ```js
+   * {
+   *   actionType: '...',
+   *   message: {...}
+   * }
+   * ```
+   * @description
+   * /!\ If wrapData is wrong configured is interpreted as false /!\
+   *
+   * @default false
+   */
+  wrapData: boolean
+
+	/**
+	 * Allow user to access to some private function to customize message reception.
+	 * This is used if you specified wrapOptions on the client side and have to manually bind message to callback.
+   * @default false
+	 */
+  exposePrivateFunction: boolean
+}
+
+export type SocketIOFileUploadClientChoose =
+{
+  files: FileList
+}
+
+export type SocketIOFileUploadClientStart<Metadata> =
+{
+  file: File & {
+    meta: Metadata
+  }
+}
+
+export type SocketIOFileUploadClientProgress<Metadata> =
+{
+  file: SocketIOFileUploadClientStart<Metadata>['file']
```

Comment:
Suggestion: does TypeScript support something like

```
export type SocketIOFileWithMetadata<Metadata> = File & { meta: Metadata }
```

It would make things cleaner.

Reply:
I agree, it's way more cleaner, I will see if this can be applied on the server side.